### PR TITLE
Dictionaries for caching Gröbner bases w.r.t. different monomial orderings

### PR DESCRIPTION
**Pull request by ederc**
After introducing `MonomialOrdering`s we can now compute and cache Gröbner bases for different orderings. This PR introduces a dictionary for handling the bases.

### CODE REVIEW

File: experimental/ModStd/ModStdQ.jl
Comment (fingolfin) [inline experimental/ModStd/ModStdQ.jl]:
```suggestion
#= TODO: Currently we had to "disable" modular GB stuff due to introducing dictionaries of GBs for ideals.
```

**Review by fingolfin — APPROVED** [review]
Fine by me. I'll leave it to @fieker to merge this once tests pass

File: experimental/ModStd/ModStdQ.jl
Comment (thofma) [inline experimental/ModStd/ModStdQ.jl]:
```suggestion
              I.gb[ord] = BiPolyArray(gd[1:length_gc], keep_ordering = false, isGB = true)
```

File: experimental/ModStd/ModStdQ.jl
Comment (thofma) [inline experimental/ModStd/ModStdQ.jl]:
```suggestion
            I.gb[degrevlex(gens(Qt))] = BiPolyArray(gd, keep_ordering = false, isGB = true)
```

File: src/Rings/MPolyQuo.jl
Comment (thofma) [inline src/Rings/MPolyQuo.jl]:
```suggestion
    G = first(values(I.gb))
```

File: src/Rings/MPolyQuo.jl
Comment (thofma) [inline src/Rings/MPolyQuo.jl]:
```suggestion
      GI = first(values(I.gb))
```

File: src/Rings/NumberField.jl
Comment (thofma) [inline src/Rings/NumberField.jl]:
```suggestion
  GI = first(values(I.gb))
```

File: src/Rings/NumberField.jl
Comment (thofma) [inline src/Rings/NumberField.jl]:
```suggestion
    GI = first(values(I.gb))
```

File: src/Rings/groebner.jl
Comment (thofma) [inline src/Rings/groebner.jl]:
```suggestion
        G = first(values(I.gb))
```

File: src/Rings/groebner.jl
Comment (thofma) [inline src/Rings/groebner.jl]:
```suggestion
    if !haskey(I.gb, ordering)
```
Could also be rewritten as
```
get!(I.gb, ordering) do _compute_groebner_basis(I.gens, ordering, complete_reduction) end
```
if you want to avoid the double lookup (should not matter here, just saying).

File: src/Rings/mpoly-affine-algebras.jl
Comment (thofma) [inline src/Rings/mpoly-affine-algebras.jl]:
```suggestion
  G = collect(first(values(J.gb)))
```

File: src/Rings/mpoly-localizations.jl
Comment (thofma) [inline src/Rings/mpoly-localizations.jl]:
```suggestion
```

File: test/runtests.jl
Comment (thofma) [inline test/runtests.jl]: ???

**Review by thofma — COMMENTED** [review]
Looks good. Just some minor comments.

**Comment by thofma** [comment]
Also pinging @HechtiDerLachs since he was asking about this a few days ago (and the code touches the localizations).

File: test/runtests.jl
Comment (ederc) [reply test/runtests.jl]: Sorry, my bad. I will reenable these tests.

**Comment by ederc** [comment]
@thofma could you please have a look at commit `3dd8f0fdc`, I tried to implement your suggestion to avoid the double lookup in `groebner_assure`. Your suggested code wasn't working, but I hope my committed variant of it is the right way to do it.

**Comment by thofma** [comment]
More like: https://github.com/oscar-system/Oscar.jl/pull/1125/commits/8c8ae0920712781b5b2453f2b40996ecfc37ea7a. (In your version, the computation of the Groebner basis was always done. Here it is only executed via an anonymous function if the key is not found.)

**Pull request merged**
